### PR TITLE
[Dynamic Buffer Calc][Mellanox] Bug fixes and enhancements for the lua plugins for buffer pool calculation and headroom checking

### PR DESCRIPTION
**Pull request by stephenxs**
**What I did**
Bug fixes for buffer pool calculation and headroom checking on Mellanox platforms.

- Test the number of lanes instead of the speed when determining whether special handling is required for a port.
  For speeds other than 400G, eg 100G, it's possible that some 100G ports have 8 lanes and others have 4 lanes,
  which means they can not share the same buffer profile.
  A suffix `_8lane` is introduced to indicate it, like `pg_lossless_100000_5m_8lane_profile`
- Take the private headroom into account when calculating the buffer pool size
- Take deviation into account when checking the headroom against the per-port limit to avoid the inaccurate result in a rare case
- Use hashtable to record the reference count of a profile in lug plugin

Signed-off-by: Stephen Sun <stephens@nvidia.com>

**Why I did it**

**How I verified it**
Run regression and manually test

**Details if related**
- Test the number of lanes instead of the speed when determining whether special handling (double headroom size) is required for a port.
  Originally, it was determined by testing whether the ports' speed is 400G but that is not accurate. A user can configure a port with 8 lanes to 100G. In this case, special handling is still required for a port that is not 400G.
  So we need to adjust the way to do that.
  The variable names are also updated accordingly: `xxx_400g` => `xxx_8lanes`
- Take deviation into account when checking the headroom against the per-port limit to avoid the inaccurate result in a rare case
  There are some deviations that make the accumulative headroom a bit larger than the quantity calculated by the buffer manager. We need to take it into account when calculating the accumulative headroom.

### CODE REVIEW

**Comment by stephenxs** [comment]
This PR contains the fixes of 2 bugs. In theory, we open 1 PR for 1 bug. But there are dependencies between both fixes, which makes it very difficult to split it.

**Comment by liat-grozovik** [comment]
@neethajohn can you please help to review?
@stephenxs to which branches this PR should be cherry picked and if this is a clean cherry picked or new new PR?

**Comment by stephenxs** [comment]
> @neethajohn can you please help to review?
> @stephenxs to which branches this PR should be cherry picked and if this is a clean cherry picked or new new PR?

202012. so far it can be smoothly cherry-picked.

File: cfgmgr/buffer_headroom_mellanox.lua [hunk 1]
```diff
@@ -26,6 +27,7 @@ local port_speed = tonumber(ARGV[1])
 local cable_length = tonumber(string.sub(ARGV[2], 1, -2))
 local port_mtu = tonumber(ARGV[3])
 local gearbox_delay = tonumber(ARGV[4])
+local is_8lane = ARGV[5]
```
Comment (neethajohn) [inline cfgmgr/buffer_headroom_mellanox.lua:30]: please rename to 'lane_cnt' and make it an 'int' so that it is generic

**Review by neethajohn — COMMENTED** [review]
Please add a sonic-mgmt tests for this 8 lane profile

**Comment by stephenxs** [comment]
> Please add a sonic-mgmt tests for this 8 lane profile

In almost every test cases, there is a logic to check whether the buffer profile in `BUFFER_PG_TABLE` is correct. On platforms with 8-lane ports, in case the speed isn't 400G the 8 lane profiles will be used and tested.
The test case PR: Azure/sonic-mgmt#3694

File: cfgmgr/buffer_headroom_mellanox.lua (hunk 1 above)
Comment (stephenxs) [reply cfgmgr/buffer_headroom_mellanox.lua]: Fixed.

Comment (stephenxs) [reply cfgmgr/buffer_headroom_mellanox.lua]: BTW, I changed to lane_count in generic code - buffermgrdyn. The lua plugin is still using is_8lane as it's Mellanox specific.

**Review by neethajohn — DISMISSED** [review]

**Review by liat-grozovik — CHANGES_REQUESTED** [review]
can you read the number of lanes and make it more generic and not Spectrum3 only?

**Comment by stephenxs** [comment]
> can you read the number of lanes and make it more generic and not Spectrum3 only?

fixed by removing the dependency on ASIC type.

File: cfgmgr/buffer_pool_mellanox.lua [hunk 2]
```diff
@@ -43,9 +40,13 @@ local function iterate_all_items(all_items)
         port = string.match(all_items[i], "Ethernet%d+")
         if port ~= nil then
             local range = string.match(all_items[i], "Ethernet%d+:([^%s]+)$")
-            local profile = redis.call('HGET', all_items[i], 'profile')
-            local index = find_profile(profile)
-            if index == 0 then
+            local profile_name = redis.call('HGET', all_items[i], 'profile')
+            if not profile_name then
+                return 1
```
Comment (liat-grozovik) [inline cfgmgr/buffer_pool_mellanox.lua:45]: as this is not a valid scenario will it be possible to add an error information for better debuggability?

File: cfgmgr/buffermgrdyn.cpp [hunk 3]
```diff
@@ -265,6 +268,24 @@ string BufferMgrDynamic::getDynamicProfileName(const string &speed, const string
         buffer_profile_key = buffer_profile_key + "_" + gearbox_model;
     }
 
+    if (m_platform == "mellanox")
+    {
+        if ((speed != "400000") && (lane_count == 8))
```
Comment (liat-grozovik) [inline cfgmgr/buffermgrdyn.cpp:273]: i am not sure i understand the logic behind this.
400g does not have a posix of 8 lanes but 100g may have but how come 100g is working with 8 lanes?

Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:273]: A user can configure a 8 lane port to 100g.

File: cfgmgr/buffer_pool_mellanox.lua (hunk 2 above)
Comment (stephenxs) [reply cfgmgr/buffer_pool_mellanox.lua:45]: We are not able to add log in lua plugins. But the result of lua will be logged by the buffer manager. By doing so, we will have info to debug.

File: cfgmgr/buffermgrdyn.cpp (hunk 3 above)
Comment (stephenxs) [reply cfgmgr/buffermgrdyn.cpp:273]: In nominal cases, 100 g port always have 4 lanes. In this case, we don’t add “8lane” to profile name.
But if user configured a 8 lanes port to 100g, we need to distinguish it from the normal case, so 8lane is added.

**Comment by stephenxs** [comment]
VS test failed due to environmental issue. Need rerun

**Comment by stephenxs** [comment]
Failed by dynamic port breakdown
```
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x50G0] FAILED [ 66%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-4x25G[10G]0] FAILED [ 66%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x50G1] FAILED [ 66%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x25G(2)+1x50G(2)0] FAILED [ 66%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x50G2] FAILED [ 67%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x50G(2)+2x25G(2)0] FAILED [ 67%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x50G3] FAILED [ 67%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x100G[40G]0] PASSED [ 67%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-4x25G[10G]1] FAILED [ 67%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x25G(2)+1x50G(2)1] FAILED [ 68%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-4x25G[10G]2] FAILED [ 68%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x50G(2)+2x25G(2)1] FAILED [ 68%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-4x25G[10G]3] FAILED [ 68%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x100G[40G]1] PASSED [ 69%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x25G(2)+1x50G(2)2] FAILED [ 69%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x50G(2)+2x25G(2)2] FAILED [ 69%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-2x25G(2)+1x50G(2)3] FAILED [ 69%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x100G[40G]2] PASSED [ 69%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x50G(2)+2x25G(2)3] FAILED [ 70%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_simple[Ethernet0-1x100G[40G]3] PASSED [ 70%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_with_vlan FAILED [ 70%]
test_port_dpb_system.py::TestPortDPBSystem::test_port_breakout_with_acl SKIPPED [ 70%]

2021-06-24T15:39:13.6879387Z         if not status:
2021-06-24T15:39:13.6879718Z             message = failure_message or (
2021-06-24T15:39:13.6883959Z                 f"Expected field/value pairs not found: expected={expected_fields}, "
2021-06-24T15:39:13.6884797Z                 f'received={result}, key="{key}", table="{table_name}"'
2021-06-24T15:39:13.6885135Z             )
2021-06-24T15:39:13.6885455Z >           assert not polling_config.strict, message
2021-06-24T15:39:13.6890298Z E           AssertionError: Expected field/value pairs not found: expected={'brkout_mode': '2x50G'}, received={'brkout_mode': '1x100G[40G]'}, key="Ethernet0", table="BREAKOUT_CFG"
2021-06-24T15:39:13.6890812Z 
2021-06-24T15:39:13.6891111Z dvslib/dvs_database.py:203: AssertionError

```

**Review by neethajohn — APPROVED** [review]

**Pull request merged**